### PR TITLE
Jdp190101 17 product testy

### PR DESCRIPTION
**Pull request by kszubra**

### CODE REVIEW

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        productRepository.save(testProductOne);
+        productRepository.save(testProductTwo);
+        productRepository.save(testProductThree);
+        productRepository.save(testProductFour);
+
+        //when
+        List<Product> loadedProducts = productRepository.findAll();
+
+        //then
+        Assert.assertEquals(4, loadedProducts.size());
```
Comment (radoslaw-dauksza) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: To samo co w pierwszym teście.

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        Product testProduct = new Product();
+        testProduct.setName("Shirt");
+        testProduct.setDescription("white shirt");
+        testProduct.setPrice(BigDecimal.valueOf(5.20));
+
+        //when
+        productRepository.save(testProduct);
+
+        //then
+        Assert.assertEquals(1L, productRepository.count());
```
Comment (radoslaw-dauksza) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: Moim zdaniem test nie będzie działał jeśli w bazie już będą zapisane jakieś produkty.

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        //given
+        Product testProduct = new Product("Shirt", "white shirt", BigDecimal.valueOf(5.20));
+        productRepository.save(testProduct);
+        Long id = testProduct.getId();
+
+        //when
+        productRepository.deleteById(id);
+
+        //then
+        Assert.assertEquals(0L, productRepository.count());
```
Comment (radoslaw-dauksza) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: W tym miejscu też będzie problem jeśli baza zawiera jakieś wpisy przed wykonaniem testu.

**Review by radoslaw-dauksza — CHANGES_REQUESTED** [review]

**Comment by lukaszmazurkiewicz** [comment]
Poza tym co wpisał @radion358 to wg mnie jest ok.

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 1 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: Poprawione :) Teraz nie sprawdza już ilości, tylko występowanie konkretnych produktów

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 2 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: Poprawione

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 3 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: Poprawione

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 4]
```diff
@@ -111,39 +112,40 @@ public void testFindAll() {
...
         Product testProduct = new Product("Shirt", "white shirt", BigDecimal.valueOf(5.20));
         productRepository.save(testProduct);
 
         //when
         productRepository.delete(testProduct);
+        List<Product> allProducts = productRepository.findAll();
 
         //then
-        Assert.assertEquals(0L, productRepository.count());
+        Assert.assertTrue( !allProducts.contains(testProduct) );
```
Comment (radoslaw-dauksza) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: nie było by lepiej assertFalse(allProducts.contains(testProduct))?

**Review by aleksanderkot00 — APPROVED** [review]
Moim zdaniem po wprowadzonych zmianach jest ok i testy u mnie przechodzą.

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 4 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: Kwestia estetyki, ale ok, zmienione

Comment (coder-msl) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: popieram.

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        Product testProduct = new Product("Shirt", "white shirt", BigDecimal.valueOf(5.20));
+        productRepository.save(testProduct);
+        Long id = testProduct.getId();
+
+        //when
+        productRepository.deleteById(id);
+        List<Product> allProducts = productRepository.findAll();
+
+        //then
+        Assert.assertTrue( !allProducts.contains(testProduct) );
```
Comment (coder-msl) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: assertFalse() tak jak powyżej sugerował @radion358 ??

**Review by radoslaw-dauksza — APPROVED** [review]

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+        //then
+        Assert.assertTrue(loadedProducts.contains(testProductOne));
+        Assert.assertTrue(loadedProducts.contains(testProductTwo));
+        Assert.assertTrue(loadedProducts.contains(testProductThree));
+        Assert.assertTrue(loadedProducts.contains(testProductFour));
+    }
+
+    @Test
+    public void testDeleteProduct() {
```
Comment (coder-msl) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:123]: Na początku napisałem zupełnie inny komentarz ale potem sprawdziłem i opatrzenie tego testu @Transactional niczego nie psuje:

2019-02-14 00:14:36.470  INFO 19977 --- [           main] o.s.t.c.transaction.TransactionContext   : Began transaction (1) for test context [DefaultTestContext@4135c3b testClass = ProductTestSuite, testInstance = com.kodilla.ecommercee.domain.ProductTestSuite@7a2a7492, testMethod = testDeleteProduct@ProductTestSuite, testException = [null], mergedContextConfiguration = [WebMergedContextConfiguration@6302bbb1 testClass = ProductTestSuite, locations = '{}', classes = '{class com.kodilla.ecommercee.EcommerceeApplication}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@52feb982, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@bd8db5a, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@4b553d26, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@2c039ac6], resourceBasePath = 'src/main/webapp', contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map['org.springframework.test.context.web.ServletTestExecutionListener.activateListener' -> true, 'org.springframework.test.context.web.ServletTestExecutionListener.populatedRequestContextHolder' -> true, 'org.springframework.test.context.web.ServletTestExecutionListener.resetRequestContextHolder' -> true]]; transaction manager [org.springframework.orm.jpa.JpaTransactionManager@1f45db49]; rollback [true]
Hibernate: select next_val as id_val from hibernate_sequence for update
Hibernate: update hibernate_sequence set next_val= ? where next_val=?
2019-02-14 00:14:36.820  INFO 19977 --- [           main] o.h.h.i.QueryTranslatorFactoryInitiator  : HHH000397: Using ASTQueryTranslatorFactory
Hibernate: insert into products (description, group_id, name, price, product_id) values (?, ?, ?, ?, ?)
Hibernate: delete from products where product_id=?
Hibernate: select product0_.product_id as product_1_4_, product0_.description as descript2_4_, product0_.group_id as group_id5_4_, product0_.name as name3_4_, product0_.price as price4_4_ from products product0_
2019-02-14 00:14:37.808  INFO 19977 --- [           main] o.s.t.c.transaction.TransactionContext   : Rolled back transaction for test:

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        productRepository.delete(testProduct);
+        List<Product> allProducts = productRepository.findAll();
+
+        //then
+        Assert.assertTrue( !allProducts.contains(testProduct) );
+
+    }
+
+    @Test
+    public void testDeleteProductById() {
```
Comment (coder-msl) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:139]: Na początku napisałem zupełnie inny komentarz ale potem sprawdziłem i opatrzenie tego testu @transactional niczego nie psuje:

2019-02-14 00:14:36.470 INFO 19977 --- [ main] o.s.t.c.transaction.TransactionContext : Began transaction (1) for test context [DefaultTestContext@4135c3b testClass = ProductTestSuite, testInstance = com.kodilla.ecommercee.domain.ProductTestSuite@7a2a7492, testMethod = testDeleteProduct@ProductTestSuite, testException = [null], mergedContextConfiguration = [WebMergedContextConfiguration@6302bbb1 testClass = ProductTestSuite, locations = '{}', classes = '{class com.kodilla.ecommercee.EcommerceeApplication}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@52feb982, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@bd8db5a, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@4b553d26, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@2c039ac6], resourceBasePath = 'src/main/webapp', contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map['org.springframework.test.context.web.ServletTestExecutionListener.activateListener' -> true, 'org.springframework.test.context.web.ServletTestExecutionListener.populatedRequestContextHolder' -> true, 'org.springframework.test.context.web.ServletTestExecutionListener.resetRequestContextHolder' -> true]]; transaction manager [org.springframework.orm.jpa.JpaTransactionManager@1f45db49]; rollback [true]
Hibernate: select next_val as id_val from hibernate_sequence for update
Hibernate: update hibernate_sequence set next_val= ? where next_val=?
2019-02-14 00:14:36.820 INFO 19977 --- [ main] o.h.h.i.QueryTranslatorFactoryInitiator : HHH000397: Using ASTQueryTranslatorFactory
Hibernate: insert into products (description, group_id, name, price, product_id) values (?, ?, ?, ?, ?)
Hibernate: delete from products where product_id=?
Hibernate: select product0_.product_id as product_1_4_, product0_.description as descript2_4_, product0_.group_id as group_id5_4_, product0_.name as name3_4_, product0_.price as price4_4_ from products product0_
2019-02-14 00:14:37.808 INFO 19977 --- [ main] o.s.t.c.transaction.TransactionContext : Rolled back transaction for test:

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        testProduct.setName("Shirt");
+        testProduct.setDescription("white shirt");
+        testProduct.setPrice(BigDecimal.valueOf(5.20));
+
+        //when
+        productRepository.save(testProduct);
+        List<Product> allProducts = productRepository.findAll();
+
+        //then
+        Assert.assertTrue(allProducts.contains(testProduct));
```
Comment (coder-msl) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:39]: W mojej ocenie przy obecnej konfiguracji H2 transakcja jest zwijana po każdym teście - o ile jest opatrzona @Transactional. Więc można było normalnie testować ilości... poprawcie mnie proszę jeśli się mylę bo jest już późno :)

2019-02-14 00:19:52.985  INFO 20154 --- [           main] o.s.t.c.transaction.TransactionContext   : Rolled back transaction for test:

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    @Autowired
+    private ProductRepository productRepository;
+    @Autowired
+    private GroupRepository groupRepository;
+
+    @Transactional
+    @Test
+    public void testSaveProduct() {
+        //given
+        Product testProduct = new Product();
```
Comment (coder-msl) [inline src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:29]: wszędzie mamy all-args-constructor użyty a w tym teście leci bezparametrowy. To celowo? Wydaje mi się, że można ujednolicić.. no chyba, że testujemy setery? :-P

**Review by coder-msl — CHANGES_REQUESTED** [review]

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 9 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:29]: Dokładnie, po "upublicznieniu" setterów chciałem w jednym teście je przetestować, tak dla pewności, że działają. Po sprawdzeniu, w pozostałych testach używałem all-argsów ponieważ były wygodniejsze niż za każdym razem korzystać z setterów, zwłąszcza przy kilku obiektach :)

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 6 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:123]: Dodałem @Transactional i też testy wykonują się poprawnie :)

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 8 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:39]: Teoretycznie tak powinno być, ale skoro zaczęliśmy już pisać w jeden, "mega bezpieczny" sposób, dlatego tak zmieniłem wedle requesta

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 5 above)
Comment (kszubra) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java]: Jest zmienione

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 9 above)
Comment (coder-msl) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:29]: w takim wypadku ten test testuje 2 rzeczy a jedną niejawnie... a powinien być unit ;) kwestia sporna czy należy testować settery... jeśli to jest typowy setter bez dodatkowej logiki - IMHO nie powinno się tego testować, a jeśli chcemy to testować to warto osobny test pod to zrobić i zrobić resztę testów jako zależnych. W sensie, że jak konstruktor nie działa poprawnie czy setter - to po co testować logikę zapisu danych, prawda? 

Na chwilę obecną OK :)

File: src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java (hunk 8 above)
Comment (coder-msl) [reply src/test/java/com/kodilla/ecommercee/domain/ProductTestSuite.java:39]: no i właśnie tu pojawia się problem. Bo w mojej ocenie polegamy na tym, że inne testy będą napisane w ten "bezpieczny" sposób. 

OK, zostawmy.

**Review by coder-msl — APPROVED** [review]

**Pull request merged**